internal/command/views/json/diagnostic.go

```diff
@@ -144,269 +145,340 @@ func NewDiagnostic(diag tfdiags.Diagnostic, sources map[string]*hcl.File) *Diagn
+	}
 
-		// Empty ranges result in odd diagnostic output, so extend the end to
-		// ensure there's at least one byte in the snippet or highlight.
-		if snippetRange.Empty() {
-			snippetRange.End.Byte++
-			snippetRange.End.Column++
-		}
-		if highlightRange.Empty() {
-			highlightRange.End.Byte++
-			highlightRange.End.Column++
-		}
+	snippetRange := highlightRange
+	if context != nil {
+		snippetRange = context.ToHCL()
+	}
 
-		diagnostic.Range = &DiagnosticRange{
-			Filename: highlightRange.Filename,
-			Start: Pos{
-				Line:   highlightRange.Start.Line,
-				Column: highlightRange.Start.Column,
-				Byte:   highlightRange.Start.Byte,
-			},
-			End: Pos{
-				Line:   highlightRange.End.Line,
-				Column: highlightRange.End.Column,
-				Byte:   highlightRange.End.Byte,
-			},
-		}
+	// Make sure the snippet includes the highlight. This should be true
+	// for any reasonable diagnostic, but we'll make sure.
+	snippetRange = hcl.RangeOver(snippetRange, highlightRange)
 
-		var src []byte
-		if sources != nil {
-			if f, ok := sources[highlightRange.Filename]; ok {
-				src = f.Bytes
-			}
-		}
+	// Empty ranges result in odd diagnostic output, so extend the end to
+	// ensure there's at least one byte in the snippet or highlight.
+	if highlightRange.Empty() {
+		highlightRange.End.Byte++
+		highlightRange.End.Column++
+	}
+	if snippetRange.Empty() {
+		snippetRange.End.Byte++
+		snippetRange.End.Column++
+	}
 
-		// If we have a source file for the diagnostic, we can emit a code
-		// snippet.
-		if src != nil {
-			diagnostic.Snippet = &DiagnosticSnippet{
-				StartLine: snippetRange.Start.Line,
+	retHighlight := tfdiags.SourceRangeFromHCL(highlightRange)
+	retSnippet := tfdiags.SourceRangeFromHCL(snippetRange)
+	return &retHighlight, &retSnippet
+}
 
-				// Ensure that the default Values struct is an empty array, as this
-				// makes consuming the JSON structure easier in most languages.
-				Values: []DiagnosticExpressionValue{},
-			}
+func newDiagnosticRange(highlightRange *tfdiags.SourceRange) *DiagnosticRange {
+	if highlightRange == nil {
+		// No particular range to report, then.
+		return nil
+	}
 
-			file, offset := parseRange(src, highlightRange)
+	return &DiagnosticRange{
+		Filename: highlightRange.Filename,
+		Start: Pos{
+			Line:   highlightRange.Start.Line,
+			Column: highlightRange.Start.Column,
+			Byte:   highlightRange.Start.Byte,
+		},
+		End: Pos{
+			Line:   highlightRange.End.Line,
+			Column: highlightRange.End.Column,
+			Byte:   highlightRange.End.Byte,
+		},
+	}
+}
 
-			// Some diagnostics may have a useful top-level context to add to
-			// the code snippet output.
-			contextStr := hcled.ContextString(file, offset-1)
-			if contextStr != "" {
-				diagnostic.Snippet.Context = &contextStr
-			}
+func newDiagnosticSnippet(snippetRange, highlightRange *tfdiags.SourceRange, sources map[string]*hcl.File) *DiagnosticSnippet {
+	if snippetRange == nil || highlightRange == nil {
```

Comment:
Original code also checks if sources != nil

Reply:
Indeed, but `sources[snippetRange.Filename]` immediately below will produce `!ok` if `sources` is `nil`, so that check is not needed with _this_ shape of the code: the same branch below deals with both "we don't have _any_ sources" and "we don't have an entry for _this particular file_".

Reply:
My java brain is leaking, I always forget that go maps are null safe :+1: